src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.SdkArchiveDiff/Archive.cs

```diff
@@ -0,0 +1,124 @@
+
+    public class TarArchive : Archive
+    {
+        private string _extractedFolder;
+
+        private TarArchive(string extractedFolder)
+        {
+            _extractedFolder = extractedFolder;
+        }
+
+        public static new async Task<TarArchive> Create(string path, CancellationToken cancellationToken = default)
+        {
+            var tmpFolder = Directory.CreateTempSubdirectory(nameof(FindArchiveDiffs));
+            using (var gzStream = File.OpenRead (path))
+            using (var gzipStream = new GZipStream (gzStream, CompressionMode.Decompress))
+            {
+                await TarFile.ExtractToDirectoryAsync(gzipStream, tmpFolder.FullName, true, cancellationToken);
+            }
+            return new TarArchive(tmpFolder.FullName);
+        }
+
+        public override bool Contains(string relativePath)
+        {
+            return File.Exists(Path.Combine(_extractedFolder, relativePath));
+        }
+
+        public override string[] GetFileNames()
+        {
+            return Directory.GetFiles(_extractedFolder, "*", SearchOption.AllDirectories).Select(f => f.Substring(_extractedFolder.Length + 1)).ToArray();
+        }
+
+        public override void Dispose()
+        {
+            if (Directory.Exists(_extractedFolder))
+                Directory.Delete(_extractedFolder, true);
+        }
+    }
+
+    public class ZipFileArchive : Archive
+    {
+        private ZipArchive _archive;
+
+        private ZipFileArchive(ZipArchive archive)
+        {
+            _archive = archive;
+        }
+
+        public static ZipFileArchive Create(string path)
+        {
+            return new ZipFileArchive(new ZipArchive(File.OpenRead(path)));
+        }
+
+        public override bool Contains(string relativePath)
+        {
+            return _archive.GetEntry(relativePath) != null;
+        }
+
+        public override string[] GetFileNames()
+        {
+            return _archive.Entries.Select(e => e.FullName).ToArray();
+        }
+
+        public override void Dispose()
+        {
+            _archive.Dispose();
+        }
+    }
+
+    public static (string Version, string Rid, string extension) GetInfoFromArchivePath(string path)
+    {
+        string extension;
+        if (path.EndsWith(".tar.gz"))
+        {
+            extension = ".tar.gz";
+        }
+        else if (path.EndsWith(".zip"))
+        {
+            extension = ".zip";
+        }
+        else
+        {
+            throw new ArgumentException($"Invalid archive extension '{path}': must end with .tar.gz or .zip");
+        }
+
+        string filename = Path.GetFileName(path)[..^extension.Length];
+        var dashDelimitedParts = filename.Split('-');
+        var (rid, versionString) = dashDelimitedParts switch
+        {
+            ["dotnet", "sdk", var first, var second, var third, var fourth] when PathWithVersions.IsVersionString(first) => (third + '-' + fourth, first + '-' + second),
```

Comment:
Given the complexity of the version numbers that show up in the set of assets produced by .NET, we created a version identifier a long time ago. It's what powers creation of the aka.ms links: https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs#L51

I'm not totally sure how well it works for paths inside an SDK, but it may be worth looking at if you start to run into corner cases with the checker.

Reply:
Thanks for the link! I copied that file to this project. I tried referencing that package, but it references a different version of MSBuild that Installer uses and that caused build issues.